packages/instrumentation-typeorm/src/typeorm.ts

```diff
@@ -99,58 +139,128 @@ export class TypeormInstrumentation extends InstrumentationBase<typeof typeorm>
+                const [sql, parameters] = this.getQueryAndParameters();
+                const mainTableName = this.getMainTableName();
+                const operation = this.expressionMap.queryType;
+
+                const connectionOptions = this?.connection?.options ?? {};
+                const attributes = {
+                    [SemanticAttributes.DB_SYSTEM]: connectionOptions.type,
+                    [SemanticAttributes.DB_USER]: connectionOptions.username,
+                    [SemanticAttributes.NET_PEER_NAME]: connectionOptions.host,
+                    [SemanticAttributes.NET_PEER_PORT]: connectionOptions.port,
+                    [SemanticAttributes.DB_NAME]: connectionOptions.database,
+                    [SemanticAttributes.DB_OPERATION]: operation,
+                    [SemanticAttributes.DB_STATEMENT]: sql,
+                    [SemanticAttributes.DB_SQL_TABLE]: mainTableName,
+                };
+                const span: Span = self.tracer.startSpan(`TypeORM ${operation} ${mainTableName}`, {
+                    kind: SpanKind.CLIENT,
+                    attributes,
+                });
+
+                let response;
+                if (self._config?.enableInternalInstrumentation) {
+                    response = await context.with(context.active(), () =>
+                        endSpan(() => original.apply(this, arguments), span)
+                    );
+                } else {
+                    response = await context.with(suppressTracing(context.active()), () =>
+                        endSpan(() => original.apply(this, arguments), span)
+                    );
+                }
+                if (self._config?.responseHook) {
+                    safeExecuteInTheMiddle(
+                        () => self._config.responseHook(span, response),
+                        (e: Error) => {
+                            if (e) diag.error('typeorm instrumentation: responseHook error', e);
+                        },
+                        true
+                    );
+                }
+                return response;
             };
         };
     }
 }
 
+const endSpan = (traced: () => any | Promise<any>, span: Span) => {
+    try {
+        const result = traced();
+        if (isPromise(result)) {
+            return Promise.resolve(result)
+                .catch((err) => {
+                    if (err) {
+                        if (typeof err === 'string') {
+                            span.setStatus({ code: SpanStatusCode.ERROR, message: err });
+                        } else {
+                            span.recordException(err);
+                            span.setStatus({ code: SpanStatusCode.ERROR, message: err?.message });
+                        }
+                    }
+                    throw err;
+                })
+                .finally(() => span.end());
+        } else {
+            span.end();
+            return result;
+        }
+    } catch (error: any) {
```

Comment:
wouldn't it already be closed and error recorded by the `catch` above in case of promise?

Reply:
It's only if it's a promise, but in other cases, I also want to record the errors